python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -399,12 +409,116 @@ def get_known_used_data(
+        if first_available_time_window is None:
+            return None
+
+        # if the first partition has not been filled
+        if first_available_time_window.start < first_filled_time_window.start:
+            return None
+
+        # there are no events for this asset after the given record
+        if status_cache_value.latest_storage_id <= record_id:
+            return first_filled_time_window.end
+
+        # get a per-partition count of the new materializations
+        new_partition_counts = self.get_materialized_partition_counts(
+            asset_key, after_cursor=record_id
+        )
+
+        total_partition_counts = self.get_materialized_partition_counts(asset_key)
+
+        # these are the partitions that did not exist before this record was created
+        net_new_partitions = {
+            partition_key
+            for partition_key, new_count in new_partition_counts.items()
+            if new_count == total_partition_counts.get(partition_key)
+        }
+
+        # there are new materializations, but they don't fill any new partitions
+        if not net_new_partitions:
+            return first_filled_time_window.end
+
+        # the oldest time window that was newly filled
+        oldest_net_new_time_window = min(
+            partitions_def.time_window_for_partition_key(partition_key)
+            for partition_key in net_new_partitions
+        )
+
+        # only factor in the oldest net new time window if it breaks the current first filled time window
+        return min(
+            oldest_net_new_time_window.start,
+            first_filled_time_window.end,
+        )
+
+    def _calculate_used_data_partitioned(
+        self,
+        asset_graph: AssetGraph,
+        asset_key: AssetKey,
+        record_id: int,
+        partitions_def: TimeWindowPartitionsDefinition,
+    ) -> Dict[AssetKey, Tuple[Optional[int], Optional[float]]]:
+        from dagster._core.storage.partition_status_cache import (
+            get_and_update_asset_status_cache_values,
+        )
+
+        # return the existing upstream data times already calculated for this record (if any)
+        known_data = self.get_known_used_data(asset_key, record_id)
+        if known_data:
+            return known_data
+
+        # this is the current state of the asset, not the state of the asset at the time of record_id
+        status_cache_value = get_and_update_asset_status_cache_values(
+            instance=self._instance,
+            asset_graph=asset_graph,
+            asset_key=asset_key,
+        )[asset_key]
+
+        partition_data_time = self._calculate_partition_data_time(
+            asset_key=asset_key,
+            record_id=record_id,
+            partitions_def=partitions_def,
+            status_cache_value=status_cache_value,
+        )
+        partition_data_timestamp = partition_data_time.timestamp() if partition_data_time else None
+        from dagster import AssetSelection
```

Comment:
we typically do these imports at the top of the function. also, since it's internal, shouldn't import from the top-level package

Reply:
whoops this was me experimenting -- forgot to clean that up